packages/react-dom/src/events/accumulateTwoPhaseListeners.js

```diff
@@ -127,6 +135,30 @@ export default function accumulateTwoPhaseListeners(
             dispatchInstances.push(node);
           }
         }
+      } else if (enableScopeAPI && node.tag === ScopeComponent) {
+        const reactScope = node.stateNode.methods;
+        const eventTypeMap = reactScopeListenerStore.get(reactScope);
+        if (eventTypeMap !== undefined) {
+          const type = ((event.type: any): DOMTopLevelEventType);
+          const listeners = eventTypeMap.get(type);
+          if (listeners !== undefined) {
+            const captureListeners = Array.from(listeners.captured);
+            const bubbleListeners = Array.from(listeners.bubbled);
+
+            for (let i = 0; i < captureListeners.length; i++) {
+              const listener = captureListeners[i];
+              const {callback} = listener;
+              dispatchListeners.unshift(callback);
+              dispatchInstances.unshift(((lastHostComponent: any): Element));
+            }
+            for (let i = 0; i < bubbleListeners.length; i++) {
+              const listener = bubbleListeners[i];
+              const {callback} = listener;
+              dispatchListeners.push(callback);
+              dispatchInstances.push(((lastHostComponent: any): Element));
```

Comment:
As far as I can tell, this is what caused the production crashes with today’s sync. This is only set if we “met” a host component in the above traversal before we “met” a Scope. Does anything actually guarantee that? We should probably be more careful with `any` casts in the production codepaths.

Reply:
Never mind, see below. I got confused between two code paths. I’m still not sure this was safe though — could the innermost instance be a TextInstance for example, followed by a Scope? I think that could still crash.

Reply:
The target can only ever be a DOM element as you can't have listeners on text nodes.

Reply:
You can't have listeners, but could it be bubbling from a text node?

Reply:
Text nodes can't have children, so there should be no case where we'd ever propagate through into or from a text node.

Reply:
No, I mean something like this:

```
<button>
  <Scope>
    {'hello'}{'world'}
  </Scope>
</button>
```

If I dispatch an event on the text node, won't it trigger this case?

Reply:
How are you dispatching the event? Whilst you can manually create an bubbling event and dispatch it directly on a text node (using the Node.prototype.dispatchEvent), this won't ever occur from an actual real-world event that the user would do.

We only store the host component when it's an actual host component instance (by checking the tag). We don't store `lastHostComponent` for text host instances, which I feel is accurate here.

Reply:
>this won't ever occur from an actual real-world event that the user would do.

I wouldn't be so sure, I remember text nodes firing events in Firefox in some cases. Not saying it should be common but may happen in production.

Reply:
Added support for HostText + test in https://github.com/facebook/react/pull/18441.